### DISCUSSION TITLE
When adding trigger on_auth_user_created, can no longer signup user

### DISCUSSION THREAD

**Discussion by AlexanderMoonBit** [supabase/supabase#6518] [Questions]
Hi there,

My signup worked fine until I included a trigger to insert a row into the profile table with id and email from the new signed up user.

Trigger code:

-- inserts a row into public.users
```
create function public.handle_new_user() 
returns trigger 
language plpgsql 
security definer set search_path = public
as $$
begin
  insert into public.profiles (id, email)
  values (new.id, new.email);
  return new;
end;
$$;

-- trigger the function every time a user is created
create trigger on_auth_user_created
  after insert on auth.users
  for each row execute procedure public.handle_new_user();
```

Supabase error

Status 500
"Database error saving new user"

**Comment by GaryAustin1** [comment]
In your public table is id type uuid?  Do you have any triggers on your profile table?  
You might also check the postgres log in the dashboard.

**Reply by AlexanderMoonBit** [reply]
Hi yes id is uuid. I actually redid it and copied it from your video https://www.youtube.com/watch?v=0N6M5BBe9AE and still gives error

**Reply by GaryAustin1** [reply]
I'm not familiar the video and I'm not a SB employee.   In general the code you have above should work.  

I have almost the exact same code working for months:
```
create function handle_new_user() returns trigger
    security definer
    language plpgsql
as
$$
begin
  insert into public.users (uid, email)
  values (new.id, new.email);
  return new;
end;
$$;
```
Are you getting any errors in the log?  
Do you happen to already have that user id in your public.profiles table?

**Reply by AlexanderMoonBit** [reply]
Hi I don't have anything in my profiles table

Getting this postgres error: "Event message
duplicate key value violates unique constraint "profile_pkey""

Interesting looking in the details it says "      "detail": "Key (id)=(22ee4e98-681e-44a0-a583-da57ef489359) already exists.","

I tested with a new random email

**Reply by GaryAustin1** [reply]
Any way you could have two trigger functions running on insert, or two triggers running the same function insert?  I believe all changes would get rolled out on error so the table would stay empty...

**Reply by AlexanderMoonBit** [reply]
Okay so I got it fixed no idea exactly what was happening. I removed everything again including my profiles table and used your function and created a new trigger. Thanks a lot for helping me out! @GaryAustin1

**Reply by lautarocolazo** [reply]
I am facing the same issue but implementing what worked for you is not doing it for me. Do you have extra details on how you solved it? Thanks in advance! @AlexanderMoonBit

**Comment by AlexanderMoonBit** [comment]
I only have one trigger and one function related to this, maybe Supabase
not deleting it properly on their side

On Fri, Apr 22, 2022 at 2:03 AM GaryAustin1 ***@***.***>
wrote:

> Any way you could have two trigger functions running both trying to
> insert? I believe all changes would get rolled out on error so the table
> would stay empty...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/supabase/supabase/discussions/6518#discussioncomment-2610693>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYVSQ6D3C7OYPYTUM6DHIMTVGGJ7ZANCNFSM5T6FAKDQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Alexander Thomsen

Founder

Moonbit
***@***.***
www.moonbit.ai

**Comment by JanChodorowski** [comment]
For me it was prefixing the table with `public` and making sure there is `security definer`.

What I am doing below is making sure the everything is dropped before creating, and upon creating new user (last command), insert new row into `user_tokens` table 

```
-- Drop the trigger if it exists
DROP TRIGGER IF EXISTS create_user_tokens_record_on_signup ON auth.users;

-- Drop the function if it exists
DROP FUNCTION IF EXISTS create_user_tokens_record_on_signup();

-- Create the function
CREATE OR REPLACE FUNCTION create_user_tokens_record_on_signup() RETURNS TRIGGER
   security definer
    language plpgsql
AS $$
BEGIN
    -- Insert a new record into the user_tokens table with the user's ID and 0 tokens
    INSERT INTO public.user_tokens (user_id, count) VALUES (NEW.id, 0);
    RETURN NEW;
END;
$$;

-- Create the trigger
CREATE TRIGGER create_user_tokens_record_on_signup
AFTER INSERT ON auth.users
FOR EACH ROW
EXECUTE PROCEDURE create_user_tokens_record_on_signup();

-- Simulate adding new user
INSERT INTO auth.users (id, email, encrypted_password, is_super_admin, created_at, updated_at)
VALUES (uuid_generate_v4(),
        CONCAT('user', SUBSTRING(MD5(RANDOM()::text), 0, 8), '@example.com'),
        'hashed_password_here',
        true,
        NOW(),
        NOW());
```

**Reply by GaryAustin1** [reply]
You don't show what your error is.
Is there anything in the auth or postgres logs of the dashboard?

**Comment by AlexanderMoonBit** [comment]
Each case can often be different. Check the logs and it will give you a
clear error. Can be anything from a field having a mismatched type etc.

Alexander Thomsen

Founder

Moonbit
***@***.***
www.moonbit.ai
